### PR TITLE
Restructures the types used for slack messaging

### PR DESCRIPTION
**Pull request by allenhumphreys**
This both simplifies and complicates the type hierarchy, but I think overall it improves code readability, especially when it gets merged with @jwilson-slalom 's changes

### CODE REVIEW

File: Sources/App/Controllers/Karma/KarmaController+Slack.swift [hunk 1]
```diff
@@ -43,59 +44,55 @@ extension KarmaController {
...
 
-                let attachments = karma.map { KarmaMessage(user: $0.id ?? "", count: $0.count).karmaAttachment() }
                 return try req.client().post(responseUrl) { beforePost in
-                    let karmaResponse = KarmaResponse(attachments: attachments)
+                    let karmaResponse = KarmaStatusResponse(forSlashCommandWithKarmaStatuses: karma)
                     try beforePost.content.encode(json: karmaResponse)
                 }
-            }
-            .catch {
+            }.catch {
```
Comment (allenhumphreys) [inline Sources/App/Controllers/Karma/KarmaController+Slack.swift]: I looked into the Vapor/NIO frameworks for guidance on this. Seems this way is the way to do it

File: Package.swift [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// swift-tools-version:4.0
```
Comment (jwilson-slalom) [inline Package.swift]: Is that all that is needed? No need to update any of the dependencies ? If so, nice !

File: Sources/App/Controllers/Karma/KarmaController+Slack.swift (hunk 1 above)
Comment (jwilson-slalom) [reply Sources/App/Controllers/Karma/KarmaController+Slack.swift]: Confused ?

File: Sources/App/Controllers/Karma/KarmaController+Slack.swift [hunk 3]
```diff
@@ -43,59 +44,55 @@ extension KarmaController {
...
-                }.catchFlatMap { _ in
-                    statusRepository.save(karma: karmaMessage.statusData())
-                }.thenThrowing { karma -> Void in
-                    let response = message.response(with: karmaMessage.slackUser(),
-                                                    attachments: [karmaMessage.slackAttachment(with: karma.count)])
-                    try slack.send(message: response)
+            statusRepository
+                .find(id: change.user)
+                .flatMap {
+                    let karmaStatus = $0 ?? KarmaStatus(id: change.user, count: 0)
```
Comment (jwilson-slalom) [inline Sources/App/Controllers/Karma/KarmaController+Slack.swift]: Oh I see what you are doing here. Nice find to clean up one block

File: Sources/App/Controllers/Karma/KarmaParser.swift [hunk 4]
```diff
@@ -38,33 +38,21 @@ final class KarmaParser {
         }
     }
 
-    private func process(match: NSTextCheckingResult, on message: String) -> KarmaMessage {
+    private func process(match: NSTextCheckingResult, on message: String) -> KarmaAdjustment {
         let groups = match.captureGroups(testedString: message)
 
         let userId = groups[0]
+        let karma = (groups[1].count - 1) * (groups[1].contains("+") ? 1 : -1)
```
Comment (jwilson-slalom) [inline Sources/App/Controllers/Karma/KarmaParser.swift]: 👌

File: Package.swift (hunk 2 above)
Comment (allenhumphreys) [reply Package.swift]: Yep!

File: Sources/App/Controllers/Karma/KarmaController+Slack.swift (hunk 3 above)
Comment (allenhumphreys) [reply Sources/App/Controllers/Karma/KarmaController+Slack.swift]: It's a little code golf-y, I'll admit! I just got inspired to go further though...

```
statusRepository.save(karma: KarmaStatus(id: change.user, count: ($0?.count ?? 0) + change.count))
```

I actually think this is more readable!

File: Sources/App/Controllers/Karma/KarmaParser.swift (hunk 4 above)
Comment (allenhumphreys) [reply Sources/App/Controllers/Karma/KarmaParser.swift]: More 🏌️ !

File: Sources/App/Slack/Slack.swift [hunk 5]
```diff
@@ -63,7 +63,7 @@ struct Slack: ServiceType {
         }
     }
 
-    func send(message: Message, onlyVisibleTo user: String) throws {
+    func send(message: SlackKitMessage, onlyVisibleTo user: String) throws {
```
[the remark this answers is not in the document]
Comment (allenhumphreys) [reply Sources/App/Slack/Slack.swift]: Definitely, I saw this last night 🙈

**Pull request merged**